packages/ubuntu_desktop_installer/lib/pages/install_alongside/install_alongside_model.dart

```diff
@@ -0,0 +1,137 @@
+
+import '../../services.dart';
+
+export 'package:subiquity_client/subiquity_client.dart' show Disk, Partition;
+
+/// View model for [InstallAlongsidePage].
+class InstallAlongsideModel extends SafeChangeNotifier {
+  /// Creates a new model with the given service.
+  InstallAlongsideModel(this._service);
+
+  final DiskStorageService _service;
+  var _storages = <GuidedStorageTargetResize>[];
+  var _disks = <String, Disk>{};
+  int? _selectedIndex;
+  int? _currentSize;
+
+  /// Detailed info of the product being installed.
+  final productInfo = ProductInfoExtractor().getProductInfo();
+
+  /// A list of existing OS installations or null if not detected.
+  List<OsProber> get existingOS => _service.existingOS ?? [];
+
+  /// Number of storages available for guided partitioning.
+  int get storageCount => _storages.length;
+
+  /// The current size of the selected storage.
+  int get currentSize => _currentSize ?? selectedStorage?.recommended ?? 0;
+
+  /// The minimum size of the selected storage.
+  int get minimumSize => selectedStorage?.minimum ?? 0;
+
+  /// The maximum size of the selected storage.
+  int get maximumSize => selectedStorage?.maximum ?? 1;
+
+  /// The total size of the selected storage.
+  int get totalSize => selectedPartition?.size ?? 0;
+
+  /// The index of the currently selected storage.
+  int? get selectedIndex => _selectedIndex;
+
+  /// The disk of the currently selected guided storage.
+  Disk? get selectedDisk => getDisk(_selectedIndex ?? -1);
+
+  /// An existing OS on the currently selected guided storage.
+  OsProber? get selectedOS => getOS(_selectedIndex ?? -1);
+
+  /// The partition of the currently selected guided storage.
+  Partition? get selectedPartition => getPartition(_selectedIndex ?? -1);
+
+  /// The currently selected guided storage.
+  GuidedStorageTargetResize? get selectedStorage {
+    return getStorage(_selectedIndex ?? -1);
+  }
+
+  /// Returns the guided storage at the given index.
+  GuidedStorageTargetResize? getStorage(int index) {
+    return _storages.elementAtOrNull(index);
+  }
+
+  /// Returns the disk of the guided storage at the given index.
+  Disk? getDisk(int index) => _disks[getStorage(index)?.diskId ?? ''];
+
+  /// Returns an existing OS on the guided storage at the given index.
+  ///
+  /// Prioritizes the OS on the specific partition of the storage at the given
+  /// index. Falls back to an OS detected on another partition of the same disk
+  /// but only if a single OS is detected.
+  OsProber? getOS(int index) {
+    return getPartition(index)?.os ??
+        getDisk(index)
+            ?.partitions
+            .whereType<Partition>()
+            .singleWhereOrNull((p) => p.os != null)
+            ?.os;
+  }
+
+  /// Returns the partition of the guided storage at the given index.
+  Partition? getPartition(int index) {
+    final number = getStorage(index)?.partitionNumber;
+    return getAllPartitions(index)?.firstOrNullWhere((p) => p.number == number);
```

Comment:
It seems impossible for the `number` to be null, but if not, then what would happen in that case?

Reply:
I added a null check just to be sure. I wonder if we could change Partition.number to non-Optional in Subiquity.